build-logic/src/main/java/org/gradle/android/plugin/AndroidVersionsPlugin.java

```diff
@@ -0,0 +1,62 @@
+package org.gradle.android.plugin;
+
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.plugins.JavaPluginExtension;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.TaskProvider;
+import org.gradle.api.tasks.testing.Test;
+
+import java.util.ArrayList;
+
+public class AndroidVersionsPlugin implements Plugin<Project> {
+
+    @Override
+    public void apply(Project project) {
+        ObjectFactory objects = project.getObjects();
+        var androidVersionsContainer = objects.domainObjectContainer(AndroidVersionDomainObject.class,
+            name -> objects.newInstance(AndroidVersionDomainObject.class, name));
+        project.getExtensions().add("androidVersions", androidVersionsContainer);
+        var generateTask = project.getTasks().register("generateVersions", GenerateVersionsTask.class);
+        generateTask.configure(t -> {
+            t.getSupportedVersions().set(androidVersionsContainer);
+            t.getVersion().set((String) project.getVersion());
+        });
+
+        var sourceSets = project.getExtensions().getByType(JavaPluginExtension.class).getSourceSets();
+        var main = sourceSets.getByName(SourceSet.MAIN_SOURCE_SET_NAME);
+        main.getOutput().dir(generateTask);
+
+        // Generate a test task for each Android version and run the tests annotated with the MultiVersionTest category
+        project.afterEvaluate(p -> {
+            var tasks = new ArrayList<String>();
+            tasks.add("test");
+            androidVersionsContainer.forEach(v -> {
+                var t = registerAndroidTestTask(p, v);
+                tasks.add(t.getName());
+            });
+            project.getTasks().register("generateGithubActionsMatrix", GithubActionsMatrixTask.class).configure(t -> t.getTasks().set(tasks));
+        });
+    }
+
+    private TaskProvider<Test> registerAndroidTestTask(Project project, AndroidVersionDomainObject androidVersion) {
+        var versionSpecificTest = project.getTasks().register(androidTestTaskName(androidVersion.getName()), Test.class);
+        versionSpecificTest.configure(t -> {
+            t.setDescription("Runs the multi-version tests for AGP " + androidVersion.getName());
+            t.setGroup("verification");
+            t.systemProperty("org.gradle.android.testVersion", androidVersion.getName());
```

Comment:
It feels weird that now we have this reusable piece of logic (that generates android test tasks) that also depends on the project containing a `SpockConfig.groovy` that sets up the categories.  Perhaps we should pull `SpockConfig.groovy` and `MultiVersionTest` into the plugin, compile them as a separate sourceset and add them to the test classpath?  Or maybe solve this problem some other way completely (for instance having a separate test source set for multiversion tests)?

Reply:
Let's pick up the conversation on this again once we have a new PR available that soley focusses on the multi-version aspect.